[draft] MD-116: Dongmen (Postconfirmations) Standards (#116)
movementlabsxyz/MIP: MD/md-n/README.md

# MD-116: Dongmen (Postconfirmations) Standards

- **Description**: Provides a set of liveness and correctness requirements for Postconfirmations protocols. 
- **Authors**: [Liam Monninger](mailto:liam@movementlabs.xyz)

> - **Approval**: :red-cross:

Reviewer: ```suggestion
- **Approval**: 
```

Reviewer: should not have a status at this point. red cross would be rejected

- **Etymology**: These standards were originally drafted in the Dongmen neighborhood of Taipei. 

## Overview

As identified in [MD-3](https://github.com/movementlabsxyz/MIP/tree/main/MD/md-3), [MD-4](https://github.com/movementlabsxyz/MIP/tree/main/MD/md-4), [MD-5](https://github.com/movementlabsxyz/MIP/tree/main/MD/md-5), [MIP-34](https://github.com/movementlabsxyz/MIP/pulls?page=2&q=is%3Apr+is%3Aopen), and [MIP-37](https://github.com/movementlabsxyz/MIP/pull/37), naive interpretations--such as MCR--of the Postconfirmations protocol fall short of modern BFT expectations. 

We summarize the shortcomings relevant to these standards as follows:

> 1. **Asynchronicity**: per [FLP](https://groups.csail.mit.edu/tds/papers/Lynch/jacm85.pdf), asynchronous voting protocols cannot achieve consensus in the presence of one or more faulty processes. Thus, Asynchronous Postconfirmations protocols, i.e., those not defining a Global Stabilization Times, are **not in fact BFT consensus protocols.** MCR, for example--due its single vote per slot without any bound--may remain in permanent disagreement. 

Reviewer: ```suggestion
1. **Asynchronicity**: per [FLP](https://groups.csail.mit.edu/tds/papers/Lynch/jacm85.pdf), asynchronous voting protocols cannot achieve consensus in the presence of one or more faulty processes. Thus, Asynchronous Postconfirmations protocols, i.e., those not defining a Global Stabilization Times, are **not in fact BFT consensus protocols.** MCR, for example--if implemented with a single vote per slot and without any bound on when the slot moves to a new committee--may remain in permanent disagreement. 
```

Reviewer: We should specificy what the requirements are for this asynchrony to hold true. We should not assume MCR remains exactly the same. You may refer to a specific commit that contains a specific version of MCR or PCP, but I am very sure that that the existing MIP will have to be updated. For example, PCP as it is implemented in `.sol`, moves to a new epoch with a new committee if no majority is reached. This means that the inital committee MAY loose its priviledge to vote on a particular slot. This, infact, renders PCP as it is implemented a synchronous protocol.

Reviewer: I would say the above is a particular problem as for MCR as there is currently no MIP

Reviewer: You may check if the branch that commited the Postconfirmation protocol describes that we have an assumption on single-vote per epoch per slot. But i would not reference the `main` branch MIP, as we are likely to update it (with brief note in the changelog)

Author: >  For example, PCP as it is implemented in .sol, moves to a new epoch with a new committee if no majority is reached. This is possible because the L1 delivers a synchronizing clock. This means that the epoch and thus the responsible committee for a given height MUST update, as the epoch (measured in L1 time) passes. This, infact, renders PCP, as it is implemented, a synchronous protocol.

I was thinking of this yesterday as well. Epoch boundaries I think should actually be excluded from this standard. The goal of both Dongmen (an maybe Ximen) is to have subdivisions of an epoch, i.e., each super block, slot, or height have some kind of synchronicity within the epoch--mainly because epochs should generally be quite long. 

Because partial synchronicity is eventual, there's not really a categorical difference between have synchronicity carry across epochs. A simple way to become partially synchronous is just to progressive unstake the lowest vote at each epoch boundary. However, I'm leaning towards updating Ximen to still exclude this sort of thing.

> 1. **Asynchronicity**: per [FLP](https://groups.csail.mit.edu/tds/papers/Lynch/jacm85.pdf), asynchronous voting protocols cannot achieve consensus in the presence of one or more faulty processes. Thus, Asynchronous Postconfirmations protocols, i.e., those not defining a Global Stabilization Times, are **not in fact BFT consensus protocols.** MCR, for example--due its single vote per slot without any bound--may remain in permanent disagreement. 

Author: I believe it is still defined in the MCR comments.

> 1. **Asynchronicity**: per [FLP](https://groups.csail.mit.edu/tds/papers/Lynch/jacm85.pdf), asynchronous voting protocols cannot achieve consensus in the presence of one or more faulty processes. Thus, Asynchronous Postconfirmations protocols, i.e., those not defining a Global Stabilization Times, are **not in fact BFT consensus protocols.** MCR, for example--due its single vote per slot without any bound--may remain in permanent disagreement. 

Reviewer: i dont understand what comments. it needs to be defined in this document.

> 2. **Liveness**: failure to come to consensus presents a liveness shortcoming. Permanent disagreement means that the network will never progress to the next accepted state. Further, we assert indefinite disagreement is unnecessary in the context of Postconfirmations. 

Reviewer: it MUST be explained how disagreements could occur with only honest participants (this seemed to be your core argument)

Reviewer: if you indeed acknowledge this in MD-116 plz refer to the specific section.

Author: > if nodes can arrive at a wrong state but think they are honest, this requires they are accounted for in the byzantine set even in pbft

This is the not the what the quoted text is referring to. It's more briefly stating that...

1. An asynchronous protocol is not a (reliable) consensus protocol. It will sometimes show agreement, but one fault ruins it. That's FLP. 
2. Because we are trying to design a consensus protocol, this unreliability actually means unliveness. Or, a non-consensus protocol is an unlive consensus protocol.

For more detailed information on these properties, see [BFT Synchronicity and Liveness](#a1-bft-synchronicity-and-liveness).

In response to these shortcomings, we request a protocol which is **fully-synchronous**, **fork-transferable**, **fork-perfect**, and **minority-aware**. The last three of these terms are introduced and defined in this MD. 

> Because the Dongmen Postconfirmations Standards request a **fully-synchronous** protocol, we assert said protocol can no longer be traditionally-BFT. That is, non-supermajority forks may exist under these standards. For a traditional BFT standard, see the [Ximen Postconfirmations Standards]().

Author: So, we should propose a glossary universally because I am used to using "fork" to refer to everything, i.e., the fork-point, the tip of the fork branch, and the entire branch. Insofar as your notion of branch includes all of the vertices along the way, this is also everything.

## Definitions

> - **Fully-synchronous**: A network and protocol assumption in which message transmission and processing are guaranteed to complete within known, fixed bounds. A protocol that assumes this model must commit or reject decisions by a deterministically computable time.

Reviewer: ```suggestion
- **Fully-synchronous**: A network and protocol assumption in which message transmission and processing are guaranteed to complete within known, fixed bounds. Messages may arrive outside of these bounds, however, they are then ignored. A protocol that assumes this model must commit or reject decisions by a deterministically computable time.
```

- **Global Stabilization Time (GST)**: The unknown point in time after which a partially synchronous network behaves synchronously. This term is critical in distinguishing traditional BFT protocols from fully-synchronous ones, which assume GST has already occurred or is always satisfied.

> - **Fork-transferable**: A property of a consensus protocol that allows consumers to transfer application state from a forked chain of consensus rounds while preserving verifiability and auditability, even in the presence of honest partitions or temporary disagreement.

Author: I'm introducing this term to describe behavior that is honest w.r.t. to some subset of participants in the original network. An "honest partition," is what would create a long-lived fork.

> - **Fork-transferable**: A property of a consensus protocol that allows consumers to transfer application state from a forked chain of consensus rounds while preserving verifiability and auditability, even in the presence of honest partitions or temporary disagreement.

Reviewer: this needs to be defined in text. only mentioning it is not an introduction

> - **Fork-perfect**: A refinement of PBFT’s intersection safety. A fork is fork-perfect if every round of consensus along the fork intersects with at least one unit of honest voting power *relative to the fork*. This ensures local safety guarantees on a per-fork basis, even in the presence of multiple non-merging branches.

Author: We can use fork or branch. The "fork-point" is IMO the more common term to refer to the actual place in the tree that contains more than one child.

> - **Fork-perfect**: A refinement of PBFT’s intersection safety. A fork is fork-perfect if every round of consensus along the fork intersects with at least one unit of honest voting power *relative to the fork*. This ensures local safety guarantees on a per-fork basis, even in the presence of multiple non-merging branches.

Reviewer: I'll refer to wolfram for the difference between fork and branch. I do acknowledge that software development has lost its way with the terminology of branch and fork :) . But here we are more concerned with the mathematical behavior and hence should stick to the mathematical definitions. 

https://mathworld.wolfram.com/Fork.html
https://mathworld.wolfram.com/Branch.html

> - **Perfect Expert**: A theoretical construct (adapted from [expert learning models](https://people.csail.mit.edu/ronitt/COURSE/S16/notes7.pdf)) which assumes the existence of a strategy (or sequence of decisions) that would have been globally optimal or correct. In consensus, identifying this expert is analogous to identifying a perfect fork. Fork-perfectness aims to localize this ideal to a given fork in the presence of asynchrony or faults.

Author: I think this is just redundant and confusing to be honest. If you know the expert, then you know the values of the "perfect fork." But, that statement kind distracts what is actually important in this setting, i.e., the ability to produce that perfectness via PBFT.

> - **Minority-aware**: A protocol property requiring formal understanding and bounded valuation of consensus decisions made by a Byzantine minority. This property mandates a clear risk model quantifying the probability and cost of such decisions under full synchrony when liveness demands finality at fixed times.

Author: If $n = \frac{|V|}{3}$ where $V$ is the set of participants, the Byzantine minority $V_b$ is some set of nodes s.t. $|V_b| < n$.

> - **Minority-aware**: A protocol property requiring formal understanding and bounded valuation of consensus decisions made by a Byzantine minority. This property mandates a clear risk model quantifying the probability and cost of such decisions under full synchrony when liveness demands finality at fixed times.

Author: Which first sentence. There's an appendix describing this in detail.

> - **Minority-aware**: A protocol property requiring formal understanding and bounded valuation of consensus decisions made by a Byzantine minority. This property mandates a clear risk model quantifying the probability and cost of such decisions under full synchrony when liveness demands finality at fixed times.

Reviewer: link to the appendix missing

> - **Minority-aware**: A protocol property requiring formal understanding and bounded valuation of consensus decisions made by a Byzantine minority. This property mandates a clear risk model quantifying the probability and cost of such decisions under full synchrony when liveness demands finality at fixed times.

Reviewer: thanks but i meant in text.

## Desiderata

### D1: Fully-synchronous

> **User journey**: Consumers of Dongmen Postconfirmations consensus can rely on agreement to be achieved by a know Global Stabilization Time w.r.t. to the confirming ledger. 

Author: Fully-synchronous still has GST, it is the fixed time you describe. That is, partially-synchronous makes guarantees about $GST + \Delta$ where $GST$ is not known. Fully-synchronous assumes that $GST = t + \Delta$ where $t$ is some known point in time.

> **User journey**: Consumers of Dongmen Postconfirmations consensus can rely on agreement to be achieved by a know Global Stabilization Time w.r.t. to the confirming ledger. 

Author: > I think this requirement is too strong. PCP is currently implemented such that it does not require that consensus is reached within certain times. 

That is deferred to Ximen...

> Because the Dongmen Postconfirmations Standards request a fully-synchronous protocol, we assert said protocol can no longer be traditionally-BFT. That is, non-supermajority forks may exist under these standards. For a traditional BFT standard, see the [Ximen Postconfirmations Standards](https://github.com/movementlabsxyz/MIP/blob/l-monninger/dongmen-standards/MD/md-n).

> **User journey**: Consumers of Dongmen Postconfirmations consensus can rely on agreement to be achieved by a know Global Stabilization Time w.r.t. to the confirming ledger. 

Reviewer: The definition of GST is specific to partial synchronous. It is not related to synchronous model. this is not a matter of personal definition but what is the definition in academia. 

when you say there is a known bound for synchronization you mean Delta. But this is not the global stabilization time. In fact since we know Delta , GST = 0, see 
> In other words, GST is a point in time after which all network messages are delivered with a delay of at most Δ

in  https://blog.trailofbits.com/2021/11/11/motivating-global-stabilization/

Further literature
https://decentralizedthoughts.github.io/2019-06-01-2019-5-31-models/
or the original paper
https://groups.csail.mit.edu/tds/papers/Lynch/jacm88.pdf

> **Justification**: A fully-synchronous protocol is a consensus protocol under FLP. It also renders predictable points in time by which consensus will be achieved, offering qualitatively optimal liveness. 

Author: FLP alludes to the possibility of consensus of synchronous protocols by its comments on the impossibility of asynchronous consensus. Some parts of the PBFT paper or another Lamport paper might be better here.

> **Justification**: A fully-synchronous protocol is a consensus protocol under FLP. It also renders predictable points in time by which consensus will be achieved, offering qualitatively optimal liveness. 

Reviewer: i agree, the correct term is probably "under BFT assumptions"

### D2: Fork-transferable

**User journey**: Consumers of Dongmen Postconfirmations consensus can safely transfer state from the ledger by following a set of standards. 

> **Justification**: The transfer of value from a BFT system to other systems is critical in the modern DLT landscape. How to do this safely is critical to both application and general ledger usage.

Author: AH, yeah, I was going to provide an example with long-lived forks. I forgot to add the appendix.

| **Knowledge of bounds**   | None                | Exist post-GST (unknown) | Known and fixed          |
| **Liveness possible?**    | No | Yes, eventually         | Yes, by bounds |

| Feature                        | Fully Synchronous            | Partially Synchronous               |
|-------------------------------|-----------------------------------|----------------------------------------|
| **Timing Assumption**         | Fixed known bounds                | Unknown bounds before GST              |
| **Liveness Guarantee**        | Always (if bounds hold)           | Only eventually (after GST)            |
| **Safety Guarantee**          | Depends on strict timing, traditional Byzantine assumption violated          | Holds even under asynchrony            |
| **Fault Tolerance**           | Assumptions must be modified to allow a non-supermajority fork to exist, hence not traditionally BFT  | Designed to tolerate Byzantine faults  |
| **Performance**               | High under tight control          | Adaptive but may have delays           | 

### A2: PBFT and Perfectness

> [PBFT](https://pmg.csail.mit.edu/papers/osdi99.pdf) introduced a protocol which ensures consecutive rounds of consensus intersect in at least one unit of honest voting power via the Generalized Pigeon Hole Principle. The transitivity of this property ensures that a chain of consensus rounds must intersect in at least one unit of honest voting power. 

Author: Under Byzantine assumptions. This is one of the core results of the PBFT paper.

> Under an [experts model](https://people.csail.mit.edu/ghaffari/AA19/AAscript.pdf?utm_source=chatgpt.com) understanding of PBFT systems, we can make an assumption that there is a perfect expert and thus that this perfect expert is identified by consensus. This, in turn, renders PBFT a lossless approximate algorithm.

Author: It's supposed to be a bit of an oxymoron. The interesting property of PBFT if you make the perfect expert assumption is that you will always find a chain of rounds containing that perfect expert--so you have no loss to your experts model.

> Under an [experts model](https://people.csail.mit.edu/ghaffari/AA19/AAscript.pdf?utm_source=chatgpt.com) understanding of PBFT systems, we can make an assumption that there is a perfect expert and thus that this perfect expert is identified by consensus. This, in turn, renders PBFT a lossless approximate algorithm.

Reviewer: this is all quite interesting, but hard to access. i suggest to explain this a bit more in text

If we consider allowing forks, however, it is initially unclear how perfectness translates. Naively, our assumption that there is globally one perfect expert means that there is one **perfect** fork. 

> We can, however, apply a reduced criterion and state that a each **fork** be comprised of a chain of consensus rounds each of which intersect in one unit of honest voting power w.r.t. to the fork itself. We call this **fork-perfectness**.

Author: This to say if traced any fork back to its root, we would always find its messages are those of at least one perfect expert. We can apply such a reduced criterion because we are not PBFT--we are saying we admit forks but we still want each of these forks to have that core "perfectness" that PBFT provides.

Preserving **fork-perfectness** reduces to ensuring the fraction of stake which decided the original fork maintains a supermajority. If this does not occur, then it is impossible to have the guaranteed intersection. 

### A3: Awareness of Minority Decisions

> The requirement of **full-synchronicity** means that if a supermajority decision is not made by Global Stabilization Time, some form of minority decision must effectively be made. This does not inherently mean that a Byzantine fraction of voting power may decide the global state. For example, if long-lived forks are preserved, there are multiple global states and eventual supermajority consistency can be achieved via a reconstitution of supermajority stake on an honest fork. In other words, while a **partially-synchronous** system renders eventual liveness and guaranteed safety, a **full-synchronous** system can render guaranteed liveness and eventual safety. 

Author: I don't think it actually says this in ["Consensus in the Presence of Partial Synchrony"](https://groups.csail.mit.edu/tds/papers/Lynch/jacm88.pdf), so if you want to remove the notion of GST from our definition of fully-synchronous models, I think we should propose it as a glossary term. 

For me, it's nice to refer to $GST = t + \Delta$ in the synchronous case because it reminds that the stabilization of fully-synchronous networks is simply enforced by the global clock--not any kind of actual state of the network.

> The requirement of **full-synchronicity** means that if a supermajority decision is not made by Global Stabilization Time, some form of minority decision must effectively be made. This does not inherently mean that a Byzantine fraction of voting power may decide the global state. For example, if long-lived forks are preserved, there are multiple global states and eventual supermajority consistency can be achieved via a reconstitution of supermajority stake on an honest fork. In other words, while a **partially-synchronous** system renders eventual liveness and guaranteed safety, a **full-synchronous** system can render guaranteed liveness and eventual safety. 

Author: I'll add another appendix.

> The requirement of **full-synchronicity** means that if a supermajority decision is not made by Global Stabilization Time, some form of minority decision must effectively be made. This does not inherently mean that a Byzantine fraction of voting power may decide the global state. For example, if long-lived forks are preserved, there are multiple global states and eventual supermajority consistency can be achieved via a reconstitution of supermajority stake on an honest fork. In other words, while a **partially-synchronous** system renders eventual liveness and guaranteed safety, a **full-synchronous** system can render guaranteed liveness and eventual safety. 

Reviewer: I will argue that per definition of GST, the GST = 0 . see my response in one of the previous comments. This is not my definition but rather a result of after GST there is a fixed bound Delta within which stuff is resolved. In synchronous there is just Delta.

Regardless, Dongmen Postconfirmation protocols are required to provide a formal model of the expected value of consensus on a Byzantine minority w.r.t. the value of the state and intrinsic rewards. In other words, they must explain the nature of eventual safety that abides by Byzantine assumptions or else describe the non-BFT nature of the protocol at full-synchronous decision points. 

### A4: Example Minority-Aware Protocol

> Consider the following fully-synchronous protocol:

Reviewer: it is not clear how the voting would work. i take it that within a given duration $d$ you can only vote once for a given height? You should explicitly state such requirements.

Author: I left this as sketch because the important thing for the MD is not the algorithm, but the way it outlines how Dongmen Protocols should describe their BFT qualities.

> 

Reviewer: it is not clear until later what your height refers to. You should explicitly say that you mean slot height

1. Votes $v \in V$ are cast for states $s \in S$ at height $h \in H$.
2. A decision in made on a vote $s, h$ by time $t \in T$.
    1. If $V_{h}(s) > \frac{2*|V|}{3}$, accept $s, h$
    2. Otherwise, begin **Play Foward** algorithm.

The **Play Forward** algorithm is as follows:

> Given a duration $d: (t + d) \in T$, and a slot duration $l: (d/l) \in T$, for each slot $l'$ in the duration $t, t + l, ... t + d$ $h' = h + (l'/l)$. 

Reviewer: There are a few errors in that line
- $l$ is not a slot duration but a resolution of the duration $d$, wheras $d/l$ seems the slot duration
- the end of the sentence is not very readable, define h' not within the equation. 
- its confusing to define l' as duration, whereas you define l as a resolution factor. Use the same type of units (i.e. time) for them. indeed it doesnt seem necessary to define a resolution.
-  also something seems wrong about the end of the sentence 
- generally this sentence is hard to read, we are not only targeting math people

Reviewer: ```suggestion
Assume durations $d$ , and a slot duration $l$, where $l<d$. Assume the protocol progresses in steps of slot duration $l$, i.e. the following time steps exist: $t, t+l , t+2l, .. , t+l' \in T$ , where $l` = \delta h \cdot l$ and $h' = h + \delta h$
```

> If $l' > d$, accept the root $s, h$ of the heaviest remaining subtree by weight. 

Reviewer: - more consistency. use (s,h) not s,h
- you have not defined what means heaviest subtree, or how this results in a weighted tree ( i am aware how this would works, but it should be written explicitly)

> Otherwise, for each tuple $(s, h') \in V_{h'}$ 

Reviewer: - V_h is not set of tuples
- the "otherwise  ... 1)" should come before the above, as it is the primary check, no?

> 1. If $V_{h'}(s) > \frac{2*|V|}{3}$, accept $s, h$, i.e., the original commitment at the root of the subtree. 

Reviewer: ```suggestion
1. If $V_{h'}(s) > \frac{2*|V|}{3}$, accept $(s, h)$, i.e., the original commitment at the root of the subtree. 
```

2. If $V_{h'}(s) = \text{argmin} V_{h'}(s)$, remove all tuples $s, h \leq V_{h'}(s)$ and their descendants from the entire tree.

> At any given round, the number of ways the Byzantine fraction $\frac{|V|}{3} - 1$ can be eliminated is $n' = (incomplete)$ out of the total number of combinations $n$. 

Reviewer: - first time of mention of round. it is not defined
- it is not N/3 -1 : the assumption is f/(3f+1),  N = 3f+1 , therefore f = (N-1)/3

Under an [experts model](https://people.csail.mit.edu/ghaffari/AA19/AAscript.pdf?utm_source=chatgpt.com) understanding of PBFT systems, we can make an assumption that there is a perfect expert and thus that this perfect expert is identified by consensus. This, in turn, renders PBFT a lossless approximate algorithm.

If we consider allowing forks, however, it is initially unclear how perfectness translates. Naively, our assumption that there is globally one perfect expert means that there is one **perfect** fork. 

We can, however, apply a reduced criterion and state that a each **fork** be comprised of a chain of consensus rounds each of which intersect in one unit of honest voting power w.r.t. to the fork itself. We call this **fork-perfectness**.

Preserving **fork-perfectness** reduces to ensuring the fraction of stake which decided the original fork maintains a supermajority. If this does not occur, then it is impossible to have the guaranteed intersection. 

### A3: Awareness of Minority Decisions

> The requirement of **full-synchronicity** means that if a supermajority decision is not made by Global Stabilization Time, some form of minority decision must effectively be made. This does not inherently mean that a Byzantine fraction of voting power may decide the global state. For example, if long-lived forks are preserved, there are multiple global states and eventual supermajority consistency can be achieved via a reconstitution of supermajority stake on an honest fork. In other words, while a **partially-synchronous** system renders eventual liveness and guaranteed safety, a **full-synchronous** system can render guaranteed liveness and eventual safety. 

Reviewer: > For example, if long-lived forks are preserved, there are multiple global states and eventual supermajority consistency can be achieved via a reconstitution of supermajority stake on an honest fork.

You are just describing a system that favors liveness over safety. in the system you describe there is never any finality. Which makes any quasi-synchronous decission void. you may as well not take a decission at point Delta.
